### PR TITLE
First draft of captain event

### PR DESCRIPTION
**Pull request by Ragnarok77-factorio**
### Brief description of the changes:
-Implemented captain event in special event instead of doing it manually.
-It seems to be working but hard to say, it needs real test live as I could only test it locally with 3 clients, maybe some edge cases or something are not handled.
-This PR doesn't handle group of players wanting to play together, it doesn't handle stats and so on but that would be for another PR, let keep it simple for now to be sure the foundation is good.

/!\ This PR should be tested multiple times on live test server before ever being merged. 

### Tested Changes:
- [x] I've tested the changes locally or with people.
- [ ] This PR was tested on a live game with lot of players.
- [ ] I've not tested the changes.

### CODE REVIEW

File: comfy_panel/special_games.lua [hunk 1]
```diff
@@ -231,6 +241,321 @@ local function reset_disabled_research(team)
...
+		},
+		
+		scale = scaleChosen,
+		font = fontChosen,
+		alignment = "center",
+		scale_with_zoom = false
+	}
+end
+
+local function generateGenericRenderingCaptain()
```
Comment (TheBigZet) [inline comfy_panel/special_games.lua:441]: Is there a reason for each line having its own LuaRendering entity?

File: comfy_panel/special_games.lua [hunk 2]
```diff
@@ -385,8 +749,132 @@ local function on_gui_click(event)
...
 		generate_disabled_entities(team, eq)
+	elseif element.name == "captain_yes_choice" then
+		table.insert(global.special_games_variables["captain_mode"]["captainList"],player.name)
+		delete_player_from_novote_list(player.name)
+		if player.gui.center["captain_poll_frame"] ~= nil then player.gui.center["captain_poll_frame"].destroy() end
+		game.print(player.name .. ' wants to become a captain ! ')
+		local listCaptains = ""
+		for _,cptPlayer in pairs(global.special_games_variables["captain_mode"]["captainList"]) do
+			listCaptains = cptPlayer .. ',' .. listCaptains
+		end
```
Comment (TheBigZet) [inline comfy_panel/special_games.lua:760]:
```suggestion
		local listCaptains = table.concat(global.special_games_variables["captain_mode"]["captainList"], " ,")	
		
```

**Comment by TheBigZet** [comment]
That's a lot to get through, I don't know if I'll be able to check it all quickly

**Comment by Ragnarok77-factorio** [comment]
> That's a lot to get through, I don't know if I'll be able to check it all quickly

Oh no worries and hurry at all ! It's an event PR anyways so quite low priority compared to perf and bugfix.

I also need to fix one thing still, just thought of it after making the PR, it will start captain once both team finished picking, whereas we want prep time for captains before starting it (minor thing to do, just one more popup to do but still need to do it).

File: comfy_panel/special_games.lua (hunk 2 above)
Comment (Ragnarok77-factorio) [reply comfy_panel/special_games.lua]: Good idea, thanks, less lines, more compact, maybe even faster, did that for other lists too.

File: comfy_panel/special_games.lua (hunk 1 above)
Comment (Ragnarok77-factorio) [reply comfy_panel/special_games.lua:433]: No idea how to do a new line to be honest (I tried <br> or <br /> or \n , none of them worked so.. @TheBigZet

**Comment by Ragnarok77-factorio** [comment]
I added prep time for captain (so now they can say when their team is ready), with the possibility for referee to force start it. 
Fixed a bug where basically captain was always the same picking (forgot a .name... when comparing), renamed some messages, removed an useless option that would cause confusion for admins

File: comfy_panel/special_games.lua (hunk 1 above)
Comment (TheBigZet) [reply comfy_panel/special_games.lua:442]: Huh

**Comment by Ragnarok77-factorio** [comment]
I added late join option for referee (he can send spectators to south or north), so now a referee should be admin independent (just need admin to start event, obviously it doesn't grant any admin power to referee, debatable if he should be able to have more power.. meh, don't think it's needed, just a team balancer, not temp admin).
I also fixed some cases I didn't notice, and made code a bit more compact, and cleaned code a bit more.

I guess over time, players/referees will have feedback and change some things about it, but that's a first version. I hope I caught all bugs too, quite painful because lot of condition because an event triggers another one and so on and so on but a case could be forgotten...

**Comment by Ragnarok77-factorio** [comment]
Here is how it is , to get a view of what is happening : 
image

**Comment by developer-8** [comment]
1. Is the referee allowed to be the captain? If yes then referee-capitan dont have gui to force teams ready event.
2. After map restarted referee has the "Send player" gui
3. I don't like the text "pick for the weaker one". If they cant decide who want to pick first add pick random button and message that fortune decided that Playername will be pick first
4. During game the usual join, autojoin buttons work. It seems they shouldn’t, it says that tournament mode is enabled and ask to join a team on the island

**Comment by Ragnarok77-factorio** [comment]
> 1. Is the referee allowed to be the captain? If yes then referee-capitan dont have gui to force teams ready event.
> 
>     2. After map restarted referee has the "Send player" gui
> 
>     3. I don't like the text "pick for the weaker one". If they cant decide who want to pick first add pick random button and message that fortune decided that Playername will be pick first
> 
>     4. During game the usual join, autojoin buttons work. It seems they shouldn’t, it says that tournament mode is enabled and ask to join a team on the island

1 => Yes he's allowed (I didn't want to limit options for players), fixed it

2 => Fixed it.

3 => Okay, updated with a third choice for first pick which is random : "The first captain that will pick will be chosen randomly"

4 => Fixed it, I don't why, I disabled tournament mode when it started, oh probably because at first I didn't want to manage late joiners but it should be managed, so reverted my choice and added late join option but forgot to change that, very good catch.

I hope I didn't break anything, did a few tests, was fine...?

Thanks again !


Oh new bugfix too, game is still in tournament mode after event restart, should be fixed now too.

**Comment by Ragnarok77-factorio** [comment]
Dammit, Max made me realized I forgot to add difficulty choice integrated in the event, will fix it later once I've a bit of time again.

**Comment by Ragnarok77-factorio** [comment]
Okay, fixed it with latest commit now, and some bugfixes too : 
The new with latest commit : Added clear corpses when event starts/is canceled, bugfix added missing reset of variables when referee not found, added switch to spectator if event canceled, added allow vote for prep phase for 3 min 
and added possibility for players to vote on tournament mode during captain event

**Comment by developer-8** [comment]
Player cant go spectate because of tournament mode. A non-admin referee cannot use the team manager to move a player to spectator island.

**Comment by Ragnarok77-factorio** [comment]
For trace :
Event today will be played with this commit for now (without the 2 fixes below, unless server is updated but I don't know if it's worth bothering serv admin just for that..Since it will be on another server, can always grant temp admin if needed for ref and both issue below can be workaround then) : [e89d737]

@developer-8   : 
-Non admin referee cant use team manager => Fixed and removed ugly popup to change team of player, why did I even bother make a custom team manager when I could just grant temporary the power for referee to use team manager switch team of player option, was dumb from me...

-Players cant go spectate due to tourny => I was thinking for a bit, is it really a good idea, but I guess if they are bored of players, they should be able to just watch instead , so fixed, added the possibility for them to spectate

**Comment by developer-8** [comment]
I mean "player can't go spectate and non-admin referee cannot move the player to spectator" as one issue. Don't have anything against if player by itself can't spectate but can with help of referee. Sorry not write it clear

**Comment by developer-8** [comment]
I understand that you have already fixed it.
Use team manager is good decision.

**Comment by Ragnarok77-factorio** [comment]
Oh it's okay, no worry !

We'll see for these ones : 
=>Fix freeze bug where no one could move anymore and stucked forever (high priority)
=>Late joiner mechanism popup maybe (send popup to all spectators and captain can pick the late players..?) 
=>Captain message are in bold (someone asked for it, to see, maybe, maybe not)
=>Someone said the idea of having only captain handling research (ugh... maybe could do a poll for that.. idk)
=>Captain could pick name of team (maybe, if there is a way to avoid any unexpected bug due to special character or whatever and easy to implement...

We tested yesterday the event, it was working well for picking phase, ready phase, start phase. Quite happy of the result. No mass desync too.

For the freeze bug, I understood how it happened.
Basically when it's bugged, you have that : 
image
The list of actions "saved" is very small => impossible to unfreeze as it will only allow these actions.
Basically it happens, if you do 2 freeze in a row, you dont have this issue with UI team manager because you can freeze twice in a row. But with captain event, it can happen ! 
This case is when an admin force freeze the game, then admin starts special event captain (which is another freeze due to no check) => 2 freeze in a row => players frozen forever. Basically just do a check before doing a freeze/unfreeze and it won't happen again I think.

Very easy to reproduce : 
start fresh game, use team manager to freeze team, start captain event, now it's frozen forever (you can just spam click popup and autojoin for exemple, or use team manager to force join a team), and you will be frozen forever, even if you try to do freeze/unfreeze or start another captain event, because list of action is now rip.

I think I fixed it in this commit : [07670a8]

**Comment by Ragnarok77-factorio** [comment]
Bugfix for votes in edge cases (hated this part as it was truth tables and probably did not do it the best way, but it works), added possibility for captains to set a team name, and possibility to start a poll for late joiners (wrote this comment because no clue if you read commit messages).

We will try this new version at the next captain event and hope I didn't introduce any new bug. 😨  There are so many edge cases it's quite a painful PR. I think though main features were implemented, don't see anything else I would need to add.

Or maybe instead of areas for players, I should do a poll instead and show it at picking phase for easier picking 🤔 But tbh, not even sure it's worth it, it's a nice QoL but areas work fine too so nope for now.

**Comment by Ragnarok77-factorio** [comment]
==
After captain event of today :
=>Inventory get wiped for when players go to spectate? (he used spectated then I switched him back and rip)
=>If referees click twice on "late join", referee gets 2 popups (and no idea for what happens to players, should be cleared)
=>I was also thinking of adding auto trust for the event but... We have to do this because test server, so probably not and use the tbz command instead
=>I think someone complained that he had 2 popups when we started captain "yes want to be cpt or not, or not playing" and I had to reset the captain special event but maybe he still had popups or not..? (to start picking phase with late joiners..?), or maybe go even deeper and if "on join" and no popup and players are saying if playing or not, shows him directly a popup, would be even better..?
=>QoL but it would be nice to have an option to say the names of players who didnt answer yet at the poll on what they are doing..?
=>Prevent captain saying team ready before vote diff is over (they should talk anyways before at least for few mins..)
=>Allow a player who spectate to come back to team..? or not and prevent to spectate..? As they could use to spy technology, or maybe hide technology if in a team..? I don't know, need some thinking about that topic.
I hope I didn't forget anything..?
=>Heard  that there was a luagui error but I think it's the case for all special events, not just cpt (tbc), a check was missing or something..?

It's mostly QoL stuff/minor stuff but it will make it easier to organize for admins doing it for first time.

**Comment by TheBigZet** [comment]
> Inventory get wiped for when players go to spectate? 

Yup, the player is reset when being switched via team manager

>I was also thinking of adding auto trust for the event but... We have to do this because test server, so probably not and use the tbz command instead

Test server has the same trustlist as the main one

>I think someone complained that he had 2 popups (...)

Maybe instead of doing popups and chat printing, make one gui openable by a top button (appearing only at the event), with all the stats, choosing, picking etc? (Idk how it looks from captain perspective now)

>Heard that there was a luagui error but I think it's the case for all special events, not just cpt (tbc), a check was missing or something..?

It's likely going to be solved in #331

**Comment by Ragnarok77-factorio** [comment]
Thanks for information. 
>>For inventory issue, since team manager changes a bit how it works, instead of changing again existing function by adding condition (which made code worse), and to be honest I think players should be able to join back their team, since they were picked they should be freely be able to spectate/join team without having a referee to handle them. So I allowed the "join back" if they had a team. So no more inventory issue.

>>Trustlist, interesting, good to know, then I don't get why players complain about trust..? Only thing I'm seeing is I would guess it's the players who doesn't have perma trust, and were granted temp trust on main server that complain, and since it's temporary, this temp trust they got on main is not transfered to test server too, which is why they complain..?


>>For the popups, I fixed it (a destroy was missing and was causing a weird issue of having sometimes popup not showing.. I hope it was this bug and not another one, but was able to reproduce that.. I'll see over time if someone complains again).

Right now, everyone get popup to say what they wanna do in game (play, be cpt, watch), and then once 2 captains are chosen, one captain gets popup to pick a player, then once done, the other captain gets popup, and so on and so on, it's no hell anymore for organizer of picking phase and it's much faster compared to handling it manually with team manager. I think players like it (got no complaint about that).

>>Good to know for the luagui spam then.

>>Another bugfix, I added a condition for spectate, players cant spectate in prepa phase (ie game didnt start yet), otherwise captains could glitch the game since the force used to transfered players when they pick is their current force, so if they go spectate => rip. No one did this bug but just in case, in case there is a weird captain.

>>Another bugfix, I clear chosen team variable now when captain events starts, otherwise we had this issue (no one did it but... you never know) : 
Start captain game, then some players are picked in a team, finally you realize a captain left or whatever, or you want to redo picking phase because 5 new players joined, so you start a new captain game, but they kept their old chosen team in variable so they could join a team whereas they must wait to be picked as usual again.

>>Captains now can't tell their team is ready if the difficulty poll is not over yet (I saw sometimes captain saying ready within 2s, not even thinking of difficulty, 0 strategy talk, that's bad. At least this way even if captains don't care, the team have at least a 3 min window to strategize

>>I added a new print for players who get picked, saying they should join voice on discord (sunday, I got a few players to join discord whereas they were totally new on bb, very surprised for them to join but it worked, didn't expect that at all).

>>There were some games after event of sunday 10/12/2022, no one reported any bug so that's good to know, no side effect detected for now.

I think I did enough changes for this week (don't want to rework/change too much at same time, even if I check my changes, I want to avoid as much as possible breaking the event), let's see how it goes for the next one.

**Comment by Ragnarok77-factorio** [comment]
Switching to draft, until no more improvements to do (to avoid merging an unfinished and having to do another PR about it)

**Comment by Ragnarok77-factorio** [comment]
From the last captain game : 
=>Testing/new feature : Need to verify or test if you can do late join while captains and players are voting for diff/strategizing

=>Cleaning : Clean some texts (like FIXME should be removed, and put it back once a system was found for that, another pr for stats & history), or join late rules are not 100% accurate (or at least say "up to referee")

=>New explicit rule : Add that captains are allowed to send their team mate back to spectator island if they don't listen or disagree with captain

=>QoL : Would be nice if new player joining gets a popup asking if they wanna play or not (on join event, add popup if new..), and it would tell everyone that he wanna play (or maybe even trigger the referee late join popup 🤔 )
=>QoL : Add possibility for players to be blacklist from late joiners polls (so they dont get spammed anymore saying no every  min)
=>QoL : if only 1 player left, picked automatically no popup for captain (they dont have to think and waste time on picking the player since they have no choice)
=>QoL : On very unbalanced player wise amount games (10v15), it would be nice if referee could have a button to just "send all players" to a team, no question asked
=>QoL : New players asking where/how to join voice chat, should add in message too on how to get it 
=>QoL : Would be nice to have pop-up at beginning for all players who answered popup, to see the result of it (like who said yes/no/spectator/didnt answer), so it's a bit more lively early game

Ignored/wont do : 
=>Some players complained "ahhhh, I have to do quickbar again..", no solution found for that, 2 different servers, not worth it bothering finding a solution

==
This PR is never ending, and we didnt even reach code review yet (it's mostly QoL stuff now but..).
==

**Comment by Ragnarok77-factorio** [comment]
For the map reveal, it's to have a more fair match (since not all players use the trick and there is no way to stop it, it's better to just reveal for all)

**Comment by Ragnarok77-factorio** [comment]
Delete map reveal, it causes massive lag early game, unplayable ! Until a better solution is found (or maybe just never and whatever, it was just an idea to get more "fair fight" but doesn't matter much.

**Comment by Ragnarok77-factorio** [comment]
Revert done, no map reveal, let's see how it goes at next captain.

**Comment by Ragnarok77-factorio** [comment]
Retrieved master updates so PR of captain have last updates from main code.

**Comment by Ragnarok77-factorio** [comment]
Some new features + 1 bugfix done (check previous commit for full list of updates)

**Comment by Ragnarok77-factorio** [comment]
Group system added it was very painful as there are lot of conditions to take into account :
1 1 1 picking system
1 2 1 1 1 picking system
what if the player leaves group, join group
what if he's spectator
what if he already has a team
what if "insert random edge case"....

removed some useless messages too.
**There needs to be improvement too as it's becoming quite spammy with all the options and so on...** Maybe show options on island instead of chat, try to find other ideas instead of a text message, show it visually or dunno...


I tested it quite hard solo with multiple clients, didn't find any bug/edge case I forgot. But who knows what will happen in a real live bb game.

**Comment by Ragnarok77-factorio** [comment]
Bugs reported : 
=>Name in middle island doesn't update when using replacereferee or replacecaptain command.

=>I think the 3 minutes delay can be bypassed, need to find the exploit/root cause

=>Group was working, but players are a bit confused by it to be honest, I think there are so many specific things/print on this event that players are info overloaded. it's explained when starting the event but since there are lot of messages printed, rip..

**Comment by Ragnarok77-factorio** [comment]
replacereferee and captain commands now update middle island text

**Comment by Ragnarok77-factorio** [comment]
Ah finally found the trick, it was after restart the bug, ticks was used instead of ticks_played, resulting in the security to wait 3 minutes being bypassed. I didn't know there were 2 kinds of tick..

**Comment by developer-8** [comment]
bug 1: 
1. select option 1 or 2 in "What do you want to do for the Captain event ?"
2. log out before poll end
3. log in after poll end
4. now you on a team list, but  on the island and can be picked as late joiner to opposite team

bug 2: 
captains gui not disappear after `/instant-map-reset`

**Comment by Ragnarok77-factorio** [comment]
bug to fix : Game was started (admin did unfroze) and it probably broke the game, needs to prevent admins to unfreeze during a captain game.

**Comment by Ragnarok77-factorio** [comment]
bug to fix 2 : captain was spectator, player got switched to spectator team because the team of captain was spectator force...

**Comment by Ragnarok77-factorio** [comment]
bug to fix 3 : during captain preparing their strategy, biters attack, need to add a security to that.

**Comment by Ragnarok77-factorio** [comment]
Closing this PR , it's clearly obsolete, I did another branch with stats, this one is obsolete (dont have all the recent updates..)
